File: guandan3-web/scripts/test-capture-logs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
捕获所有控制台消息
"""
import sys
import io
from playwright.sync_api import sync_playwright
import time

if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')

print("=" * 70)
print("游戏状态检查 - 详细日志")
print("=" * 70)

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    page = browser.new_page(viewport={'width': 1920, 'height': 1080})

    all_messages = []
    def on_console(msg):
        text = msg.text
        all_messages.append({"type": msg.type, "text": text})
        # 打印所有日志
        print(f"[{msg.type}] {text}")

    page.on("console", on_console)

    print("\n启动游戏...")
    page.goto("http://localhost:3000", wait_until="networkidle", timeout=60000)
    time.sleep(2)

    button = page.query_selector('[data-testid="home-practice"]')
    if button:
        print("\n点击按钮...")
        button.click()
    else:
        print("未找到按钮")
        browser.close()
        exit(1)

    print("\n等待30秒观察日志...")
    time.sleep(30)

    print("\n" + "=" * 70)
    print("AutoStart 相关日志:")
    for msg in all_messages:
        if 'AutoStart' in msg['text'] or 'auto' in msg['text'].lower():
            print(f"  {msg['text']}")

    print("\n当前状态:")
    path = page.evaluate("() => window.location.pathname")
    print(f"  路径: {path}")

    hand_element = page.query_selector('[data-testid="room-hand"]')
    if hand_element:
        cards = hand_element.query_selector_all('[data-card-id]')
        print(f"  手牌: {len(cards)} 张")
    else:
        print(f"  手牌: 无")

    play_button = page.query_selector('[data-testid="room-play"]')
    if play_button:
        is_disabled = play_button.get_attribute('disabled') is not None
        print(f"  出牌按钮: {'禁用' if is_disabled else '可用'}")

    browser.close()

File: guandan3-web/test-capture-logs.mjs
/**
 * 捕获游戏控制台日志
 * 专注收集所有错误和警告
 */

import { chromium } from 'playwright';
import { writeFile } from 'fs/promises';

const LOGS = {
  errors: [],
  warnings: [],
  info: [],
  pageErrors: [],
  networkErrors: [],
  timestamps: []
};

function addLog(type, data) {
  const entry = { ...data, timestamp: new Date().toISOString() };
  LOGS.timestamps.push(entry.timestamp);

  switch(type) {
    case 'error':
      LOGS.errors.push(entry);
      console.error(`[ERROR] ${data.message || data.text}`);
      break;
    case 'warning':
      LOGS.warnings.push(entry);
      console.warn(`[WARN] ${data.message || data.text}`);
      break;
    case 'pageError':
      LOGS.pageErrors.push(entry);
      console.error(`[PAGE ERROR] ${data.message}`);
      break;
    case 'networkError':
      LOGS.networkErrors.push(entry);
      console.error(`[NET ERROR] ${data.url} - ${data.error}`);
      break;
    default:
      LOGS.info.push(entry);
      console.log(`[INFO] ${data.message || data.text}`);
  }
}

async function captureLogs() {
  const browser = await chromium.launch({
    headless: false,
    slowMo: 50
  });

  const context = await browser.newContext({
    viewport: { width: 1920, height: 1080 }
  });

  const page = await context.newPage();

  // 捕获控制台消息
  page.on('console', msg => {
    const type = msg.type();
    const text = msg.text();
    const location = msg.location();

    if (type === 'error') {
      addLog('error', { text, url: location?.url, line: location?.lineNumber });
    } else if (type === 'warning') {
      addLog('warning', { text, url: location?.url, line: location?.lineNumber });
    } else if (text.includes('ERROR') || text.includes('错误')) {
      // 排除正常的调试日志
      if (!text.includes('[fetchGame]') && !text.includes('gameError: null')) {
        addLog('error', { text, url: location?.url, line: location?.lineNumber });
      }
    }
  });

  // 捕获页面错误
  page.on('pageerror', error => {
    addLog('pageError', { message: error.message, stack: error.stack });
  });

  // 捕获网络错误
  page.on('requestfailed', request => {
    const failure = request.failure();
    if (failure && failure.text !== 'net::ERR_ABORTED') {
      addLog('networkError', { url: request.url(), error: failure.text });
    }
  });

  // 捕获 400+ 响应
  page.on('response', response => {
    if (response.status() >= 400) {
      addLog('networkError', { url: response.url(), error: `HTTP ${response.status()}` });
    }
  });

  try {
    console.log('=== 开始捕获游戏日志 ===\n');

    // 访问首页
    console.log('[1/5] 访问首页...');
    await page.goto('http://localhost:3000', { waitUntil: 'networkidle', timeout: 30000 });
    await page.waitForTimeout(3000);

    // 点击练习按钮
    console.log('[2/5] 进入练习房间...');
    const practiceBtn = page.locator('[data-testid="home-practice"]').first();
    if (await practiceBtn.isVisible({ timeout: 5000 })) {
      await practiceBtn.click();
      await page.waitForTimeout(5000);
    }

    // 等待游戏加载
    console.log('[3/5] 等待游戏加载...');
    await page.waitForTimeout(5000);

    // 检查开始按钮
    const startBtn = page.locator('button:has-text("开始")').first();
    if (await startBtn.isVisible({ timeout: 3000 })) {
      console.log('[4/5] 点击开始按钮...');
      await startBtn.click();
      await page.waitForTimeout(5000);
    } else {
      console.log('[4/5] 游戏可能已自动开始');
    }

    // 观察游戏运行 30 秒
    console.log('[5/5] 观察游戏运行（30秒）...');
    for (let i = 0; i < 6; i++) {
      await page.waitForTimeout(5000);
      console.log(`  观察 ${i + 1}/6...`);
    }

    // 最终截图
    await page.screenshot({ path: 'test-results/game-state-capture.png' });
    console.log('\n截图已保存');

  } finally {
    await browser.close();
  }

  // 生成报告
  const report = {
    summary: {
      totalErrors: LOGS.errors.length,
      totalWarnings: LOGS.warnings.length,
      totalPageErrors: LOGS.pageErrors.length,
      totalNetworkErrors: LOGS.networkErrors.length,
      totalIssues: LOGS.errors.length + LOGS.warnings.length + LOGS.pageErrors.length + LOGS.networkErrors.length,
      timestamp: new Date().toISOString()
    },
    errors: LOGS.errors,
    warnings: LOGS.warnings,
    pageErrors: LOGS.pageErrors,
    networkErrors: LOGS.networkErrors
  };

  await writeFile('test-results/captured-logs.json', JSON.stringify(report, null, 2));

  console.log('\n=== 日志捕获完成 ===');
  console.log(`错误: ${report.summary.totalErrors}`);
  console.log(`警告: ${report.summary.totalWarnings}`);
  console.log(`页面错误: ${report.summary.totalPageErrors}`);
  console.log(`网络错误: ${report.summary.totalNetworkErrors}`);
  console.log(`总问题数: ${report.summary.totalIssues}`);
  console.log('\n详细报告: test-results/captured-logs.json');

  if (report.summary.totalIssues > 0) {
    console.log('\n=== 问题摘要 ===');
    [...LOGS.errors, ...LOGS.pageErrors].forEach((err, i) => {
      console.log(`${i + 1}. ${err.text || err.message}`);
    });
  }

  return report;
}

captureLogs().then(report => {
  process.exit(report.summary.totalIssues > 0 ? 1 : 0);
}).catch(err => {
  console.error('测试失败:', err);
  process.exit(1);
});
